src/Compiler/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptor.cs

```diff
@@ -50,16 +51,21 @@ protected BoundAttributeDescriptor(string kind)
 
     public bool CaseSensitive { get; protected set; }
 
-    private bool? _isDirectiveAttribute;
+    private bool _directiveAttributeSet;
+    private bool _isDirectiveAttribute;
 
     public bool IsDirectiveAttribute
     {
         get
         {
-            _isDirectiveAttribute ??= Metadata.TryGetValue(ComponentMetadata.Common.DirectiveAttribute, out var value) &&
-                    string.Equals(bool.TrueString, value);
-
-            return _isDirectiveAttribute.Value;
+            if (!_directiveAttributeSet)
+            {
+                _isDirectiveAttribute = Metadata.TryGetValue(ComponentMetadata.Common.DirectiveAttribute, out var value) &&
+                        string.Equals(bool.TrueString, value);
+                _directiveAttributeSet = true;
```

Comment:
This is not threadsafe, as the write to `_directiveAttributeSet` can be reordered before the write to `_isDirectiveAttribute` (because there is no data dependency between the two). Let's make this very simple and use an `int` field, with 3 possible values: 0, 1 (false), or 2 (true).

Reply:
This is why Roslyn has a `ThreeState` enum: https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/InternalUtilities/ThreeState.cs